drivers/video/gc2145.c

```diff
@@ -0,0 +1,1150 @@
+}
+
+static int gc2145_set_window(const struct device *dev, uint16_t reg, uint16_t x, uint16_t y,
+			     uint16_t w, uint16_t h)
+{
+	int ret;
+	const struct gc2145_config *cfg = dev->config;
+
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_RESET, GC2145_SET_P0_REGS);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Y/row offset */
+	ret = gc2145_write_reg(&cfg->i2c, reg++, y >> 8);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gc2145_write_reg(&cfg->i2c, reg++, y & 0xff);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* X/col offset */
+	ret = gc2145_write_reg(&cfg->i2c, reg++, x >> 8);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gc2145_write_reg(&cfg->i2c, reg++, x & 0xff);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Window height */
+	ret = gc2145_write_reg(&cfg->i2c, reg++, h >> 8);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gc2145_write_reg(&cfg->i2c, reg++, h & 0xff);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Window width */
+	ret = gc2145_write_reg(&cfg->i2c, reg++, w >> 8);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gc2145_write_reg(&cfg->i2c, reg++, w & 0xff);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return ret;
+}
+
+static int gc2145_set_output_format(const struct device *dev, int output_format)
+{
+	int ret;
+	const struct gc2145_config *cfg = dev->config;
+
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_RESET, GC2145_SET_P0_REGS);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (output_format != VIDEO_PIX_FMT_RGB565) {
+		LOG_ERR("Image format not supported");
+		return -ENOTSUP;
+	}
+
+	/* Disable JPEG compression and set output to RGB565 */
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_OUTPUT_FMT, GC2145_REG_OUTPUT_FMT_RGB565);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_msleep(30);
+
+	return 0;
+}
+
+static int gc2145_set_resolution(const struct device *dev, uint16_t img_width, uint16_t img_height)
+{
+	int ret;
+	const struct gc2145_config *cfg = dev->config;
+
+	uint16_t win_w = UXGA_HSIZE + 16;
+	uint16_t win_h = UXGA_VSIZE + 8;
+	uint16_t c_ratio = 1;
+	uint16_t r_ratio = 1;
+	uint16_t x = (((win_w / c_ratio) - UXGA_HSIZE) / 2);
+	uint16_t y = (((win_h / r_ratio) - UXGA_VSIZE) / 2);
+	uint16_t win_x = ((UXGA_HSIZE - win_w) / 2);
+	uint16_t win_y = ((UXGA_VSIZE - win_h) / 2);
+
+	/* Set readout window first. */
+	ret = gc2145_set_window(dev, GC2145_REG_BLANK_WINDOW_BASE, win_x, win_y, win_w, win_h);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set cropping window next. */
+	ret = gc2145_set_window(dev, GC2145_REG_WINDOW_BASE, x, y, 1, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Enable crop */
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_CROP_ENABLE, GC2145_CROP_SET_ENABLE);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set Sub-sampling ratio and mode */
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_SUBSAMPLE, ((r_ratio << 4) | c_ratio));
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gc2145_write_reg(&cfg->i2c, GC2145_REG_SUBSAMPLE_MODE, GC2145_SUBSAMPLE_MODE_SMOOTH);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_msleep(30);
```

Comment:
Same here.

Reply:
This is a bit obscurous point, because those delays are typically recommended by the manufacturer (same for omnivision parts), but they are never documented on the datasheet, the maanufacturer on the other hand says that those delays are a safe way that the data arrive to the I2C, passes to the internal DSP, and give sufficient amount of time to handle the changes before the next instruction.

For now I kept them even if removing the delays seem to not broke the system, since the controls where those delays exist are slowest ones and don't seem to waste CPU (not at least for capturing images point of view), but let me know if you are strict against these delays and I remove them for this initial support (until see some issue, maybe?)

Reply:
One use-case for short delays is to have systems wake-up, do some action, and go to deep sleep mode (few uA, maybe nA power usage only) where everything is turned off.

How about a comment summarizing your insights on this? This would greatly help anyone wanting to optimize boot time using your work. :)

Reply:
```suggestion
	k_sleep(K_MSEC(30));
```